# A3: Word Embeddings and Language Modelling

Adam Ek

The lab is an exploration and learning exercise to be done in a group and also in discussion with the teachers and other students.

Write all your answers and the code in the appropriate boxes below.

In this lab we will explore constructing *static* word embeddings (i.e. word2vec) and building language models. We'll also evaluate these systems on intermediate tasks, namely word similarity and identifying "good" and "bad" sentences.

* For this we'll use pytorch. Some basic operations that will be useful can be found here: https://jhui.github.io/2018/02/09/PyTorch-Basic-operations
* In general: we are not interested in getting state-of-the-art performance :) focus on the implementation and not results of your model. For this reason, you can use a subset of the dataset: the first 5000-10 000 sentences or so, on linux/mac: ```head -n 10000 inputfile > outputfile```. 
* If possible, use the MLTGpu, it will make everything faster :)

In [82]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np

# for gpu, replace "cpu" with "cuda:n" where n is the index of the GPU
device = torch.device('cpu')

In [83]:
from torch.utils.data import Dataset, DataLoader
from collections import Counter, OrderedDict

In [84]:
import pickle

# Word2Vec embeddings

In this first part we'll construct a word2vec model which will give us *static* word embeddings (that is, they are fixed after training).

After we've trained our model we will evaluate the embeddings obtained on a word similarity task.

## Formatting data


First we need to load some data, you can download the file on Canvas under files/assignments/03-lab-data/wiki-corpus.50000.txt. The file contains 50 000 sentences randomly selected from the complete wikipedia. Each line in the file contains one sentence. The sentences are whitespace tokenized.

Your first task is to create a dataset suitable for word2vec. That is, we define some ```window_size``` then iterate over all sentences in the dataset, putting the center word in one field and the context words in another (separate the fields with ```tab```).

For example, the sentece "this is a lab" with ```window size = 4``` will be formatted as:
```
center, context
---------------------
this    is a
is      this a lab
a       this is lab
lab     is a
```

this will be our training examples when training the word2vec model.

[3 marks]

In [20]:
data_path = '/Users/tomsodahlbladsjo/mlt/compSem/computational-semantics/assignments/03-distributed-representations-and-language-models/03-lab-data/wiki-corpus.50000.txt'
WINDOW_SIZE = 4

def corpus_reader(data_path):
    with open(data_path) as f:
        dataset = []
        wz = WINDOW_SIZE  
        for line in f:
            words = line.split()
            for i in range(len(words)):
                center = words[i]
                if i < wz:
                    context = words[0:i]+words[i+1:i+wz+1]
                else:
                    context = words[i-wz:i]+words[i+1:i+wz+1]
                dataset.append(center + '\t' + ' '.join(context))
    return dataset
                
data = corpus_reader(data_path)

We sampled 50 000 senteces completely random from the *whole* wikipedia for our training data. Give some reasons why this is good, and why it might be bad. (*note*: We'll have a few questions like these, one or two reasons for and against is sufficient)

[2 marks]

#### Answer:

It is good, because then we (hopefully) get more general knowledge of the distribution of words in English (assuming that all of Wikipedia means all of *English* Wikipedia, which seems to be the case). If we only sampled from a few chosen Wikipedia articles, the knowledge might not be as easily generalizable. On the other hand, this also mean that we are adding a lot of very specific vocabulary items, such as technical terms relating to various obscure subjects, and that might result in an unnecessary large vocabulary with lots of low-frequency words (this could be solved by excluding the least common words from the vocabulary, which we considered but didn't have time to implement).


### Loading the data

We now need to load the data in an appropriate format for torchtext (https://torchtext.readthedocs.io/en/latest/). We'll use PyText for this and it'll follow the same structure as I showed you in the lecture (remember to lower-case all tokens). Create a function which returns a (bucket)iterator of the training data, and the vocabulary object (```Field```). 

(*hint1*: you can format the data such that the center word always is first, then you only need to use one field)

(*hint2*: the code I showed you during the leture is available in /files/pytorch_tutorial/ on canvas)

[4 marks]

In [89]:
class Vocab:
    def __init__(self, tokens, pad_token):
        self.tokens = tokens
        self.i2t = {i: t for i, t in enumerate(tokens, start=1)}
        self.i2t[0] = pad_token
        self.t2i = {v: k for k, v in self.i2t.items()}
        
    def __len__(self):
        return len(self.tokens)
    
    def __getitem__(self, x):
        if type(x) == str:
            return self.t2i[x]
        if type(x) == int:
            return self.i2t[x]

In [90]:
class CBOWDataset(Dataset):
    def __init__(self, data,
                 window_size=4,
                 unk_token='UNK',
                 pad_token='PAD'):
        
        self.window_size = window_size
        self.unk_token = unk_token
        self.pad_token = pad_token
        
        self.text_data = [line.lower().split('\t') for line in data]
        self.vocab = self.create_vocab()
        self.samples = self.create_samples()
        
        self.pad_index = self.vocab.t2i[self.pad_token]
        self.unk_index = self.vocab.t2i[self.unk_token]

    def create_vocab(self):
        tokens = set([self.unk_token]+[line[0] for line in self.text_data])
        vocab = Vocab(tokens, self.pad_token)
        return vocab
    
    def word2index(self,line):
        target, context = line
        t_index = self.vocab.t2i[target]
        c_indices = [self.vocab.t2i[word] for word in context.split()]
        padding = [self.vocab.t2i[self.pad_token]]*(self.window_size*2-len(c_indices)) 
        c_indices += padding
        return t_index, c_indices
    
    def create_samples(self):
        return [self.word2index(line) for line in self.text_data]
    
    
    def __getitem__(self, idx):
        
        return self.samples[idx]

    def __len__(self):
        return len(self.samples)
    

In [91]:
def cbow_get_data(data, batch_size):
    dataset = CBOWDataset(data)
    dataloader = DataLoader(dataset,
                            batch_size=batch_size,
                            shuffle=True,
                            collate_fn=lambda x: x)
    return dataloader, dataset.vocab

In [39]:
# Check that it is working and the format is right:
dataloader, vocab = cbow_get_data(data, 4)
print('pad_index =', vocab.t2i['PAD'])
for i, batch in enumerate(dataloader):
    print('Batch size:', len(batch)) 
    for example in batch:
        print(example)
  # Only look at first batch.
    break

pad_index = 0
Batch size: 4
(59824, [65226, 16603, 26652, 10350, 17090, 4967, 38929, 70508])
(69572, [4967, 52051, 36621, 45934, 67040, 41255, 746, 21260])
(24472, [59587, 58937, 49299, 70036, 48215, 0, 0, 0])
(4967, [25308, 13882, 57260, 9463, 22151, 50866, 13882, 76956])


We lower-cased all tokens above; give some reasons why this is a good idea, and why it may be harmful to our embeddings.

[2 marks]

#### Answer:
Lower-casing all tokens means that a word will not be considered a two different words when it occurs in sentence-initial position as opposed to anywhere else (probably good). This lets us capture the fact that it encodes the same semantic meaning regardless of where in the sentence it occurs, and it also keeps us from unnecessarily expanding our vocabulary with two entries (capitalized and lower-cased) for each word.
On the other hand, lower-casing every token also means that there will be no difference between proper names and other words that happen to be spelled the same, eg the word bill (as in a dollar bill) and the proper name Bill (probably bad, since they are different words with very different meanings).

## Word Embeddings Model

We will implement the CBOW model for constructing word embedding models.

In [92]:
import torch.optim as optim

In the CBOW model we try to predict the center word based on the context. That is, we take as input ```n``` context words, encode them as vectors, then combine them by summation. This will give us one embedding. We then use this embedding to predict *which* word in our vocabuary is the most likely center word. 

Implement this model 

[7 marks]

In [93]:
class CBOWModel(nn.Module):
    def __init__(self, vocab_dims, embedding_dims, pad_index):
        super(CBOWModel, self).__init__()
        self.embeddings = nn.Embedding(num_embeddings=vocab_dims, 
                                               embedding_dim=embedding_dims,
                                               padding_idx=pad_index)
        
        self.prediction = torch.nn.Linear(in_features=embedding_dims,
                                             out_features=vocab_dims, bias=False)
        
    
    def forward(self, context):
        embedded_context = self.embeddings(context)
        projection = self.projection_function(embedded_context)
        predictions = self.prediction(projection)
        
        return predictions
        
    def projection_function(self, xs):
        """
        This function will take as input a tensor of size (B, S, D)
        where B is the batch_size, S the window size, and D the dimensionality of embeddings
        this function should compute the sum over the embedding dimensions of the input, 
        that is, we transform (B, S, D) to (B, 1, D) or (B, D) 
        """
        xs_sum = xs.sum(dim=-2)
        return xs_sum

Now we need to train the models. First we define which hyperparameters to use. (You can change these, for example when *developing* your model you can use a batch size of 2 and a very low dimensionality (say 10), just to speed things up). When actually training your model *fo real*, you can use a batch size of [8,16,32,64], and embedding dimensionality of [128,256].

In [42]:
# you can change these numbers to suit your needs :)
word_embeddings_hyperparameters = {'epochs':1,
                                   'batch_size':16,
                                   'embedding_size':128,
                                   'learning_rate':0.001,
                                   'embedding_dim':10}

Train your model. Iterate over the dataset, get outputs from your model, calculate loss and backpropagate.

We mentioned in the lecture that we use Negative Log Likelihood (https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html) loss to train Word2Vec model. In this lab we'll take a shortcut when *training* and use Cross Entropy Loss (https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), basically it combines ```log_softmax``` and ```NLLLoss```. So what your model should output is a *score* for each word in our vocabulary. The ```CrossEntropyLoss``` will then assign probabilities and calculate the negative log likelihood loss.

[3 marks]

In [43]:
# load data
dataset, vocab = cbow_get_data(data, word_embeddings_hyperparameters['batch_size'])

In [56]:
%%time

# build model and construct loss/optimizer
cbow_model = CBOWModel(len(vocab.tokens)+1, word_embeddings_hyperparameters['embedding_dim'], vocab.t2i['PAD'])
cbow_model.to(device)
loss_fn = nn.CrossEntropyLoss()
loss_fn.to(device)
optimizer = optim.Adam(cbow_model.parameters(), lr=word_embeddings_hyperparameters['learning_rate'])

# start training loop
total_loss = 0
for epoch in range(word_embeddings_hyperparameters['epochs']):
    print('Epoch:', epoch+1)
    for i, batch in enumerate(dataset):
        
        context = torch.Tensor([sample[1] for sample in batch]).long().to(device)
        target_word = torch.Tensor([sample[0] for sample in batch]).to(device)
        
        # send your batch of sentences to the model
        output = cbow_model(context)
        
        # compute the loss, you'll need to reshape the input
        # you can read more about this is the documentation for
        # CrossEntropyLoss
        loss = loss_fn(output, target_word.long())
        total_loss += loss.item()
        
        # print average loss for the epoch
        print(total_loss/(i+1), end='\r') 
        
        # compute gradients
        loss.backward()
        
        # update parameters
        optimizer.step()
        
        # reset gradients
        optimizer.zero_grad()
    print()
        

Epoch: 1


KeyboardInterrupt: 

In [57]:
# Pickle trained model:
pickle.dump(cbow_model, open('cbow_model.pickle', 'wb'))

In [94]:
# Load pickled model:
cbow_model = pickle.load(open('cbow_model.pickle', 'rb'))

## Evaluating the model

We will evaluate the model on a dataset of word similarities, WordSim353 (http://alfonseca.org/eng/research/wordsim353.html , also avalable on Canvas under files/assignments/03-lab-data). The first thing we need to do is read the dataset and translate it to integers. What we'll do is to reuse the ```Field``` that records word indexes (the second output of ```get_data()```) and use it to parse the file.

The wordsim data is structured as follows:

```
word1 word2 score
...
```


The ```Field``` we got from ```read_data()``` has two built-in functions, ```stoi``` which maps a string to an integer and ```itos``` which maps an integer to a string. 

What our datareader needs to do is: 

```
for line in file:
    word1, word2, score = file.split()
    # encode word1 and word2 as integers
    word1_idx = vocab.vocab.stoi[word1]
    word2_idx = vocab.vocab.stoi[word2]
```

when we have the integers for ```word_1``` and ```word2``` we'll compute the similarity between their word embeddings with *cosine simlarity*. We can obtain the embeddings by querying the embedding layer of the model.

We calculate the cosine similarity for each word pair in the dataset, then compute the pearson correlation between the similarities we obtained with the scores given in the dataset. 

[4 marks]

In [103]:
# your code goes here

def read_wordsim(path, vocab, embeddings):
    dataset_sims = []
    model_sims = []
    word_pairs = []
    with open(path) as f:
        for line in f:
            word1, word2, score = line.split()
            if word1 in vocab.tokens and word2 in vocab.tokens:
                
                word_pairs.append((word1, word2))
                
                score = float(score)
                dataset_sims.append(score)
            
                # get the index for the word
                word1_idx = vocab.t2i[word1]
                word2_idx = vocab.t2i[word2]
            
            
                # get the embedding of the word
                word1_emb = embeddings[word1_idx]
                word2_emb = embeddings[word2_idx]
            
                # compute cosine similarity, we'll use the version included in pytorch functional
                # https://pytorch.org/docs/master/generated/torch.nn.functional.cosine_similarity.html
                cosine_similarity = F.cosine_similarity(word1_emb, word2_emb, dim=0)
            
                model_sims.append(cosine_similarity)
    
    return dataset_sims, model_sims, word_pairs

path = '/Users/tomsodahlbladsjo/mlt/compSem/computational-semantics/assignments/03-distributed-representations-and-language-models/03-lab-data/wordsim_similarity_goldstandard.txt'
embeddings = cbow_model.embeddings.weight.detach().cpu()
data, model, word_pairs = read_wordsim(path, vocab, embeddings)
pearson_correlation = np.corrcoef(data, model)
            
# the non-diagonals give the pearson correlation
print(pearson_correlation)

[[1.         0.02589834]
 [0.02589834 1.        ]]


Do you think the model performs good or bad? Why?

[3 marks]

#### Answer:
Since the Pearson correlation coefficient ranges between -1 and 1, a value above 0 is at least alright, but of course, it could have been better. 

Select the 10 best and 10 worst performing word pairs, can you see any patterns that explain why *these* are the best and worst word pairs?

[3 marks]

Suggest some ways of improving the model we apply to WordSim353.

[3 marks]

#### Answer:
First of all, longer training would probably help since the loss seemed to still be going down at the end of training, but with the memory issues on the gpu and the slowness of training on cpu, this wasn't really feasible.
Testing different hyperparameters might also help. For example, we trained with only 10 embedding dimensions to save time, which might not have captured enough information.

If we consider a scenario where we use these embeddings in a downstream task, for example sentiment analysis (roughly: determining whether a sentence is positive or negative). 

Give some examples why the sentiment analysis model would benefit from our embeddnings and one examples why our embeddings could hur the performance of the sentiment model.

[3 marks]

# Language modeling

In this second part we'll build a simple LSTM language model. Your task is to construct a model which takes a sentence as input and predict the next word for each word in the sentence. For this you'll use the ```LSTM``` class provided by PyTorch (https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html). You can read more about the LSTM here: https://colah.github.io/posts/2015-08-Understanding-LSTMs/

NOTE!!!: Use the same dataset (wiki-corpus.50000.txt) as before.

Our setup is similar to before, we first encode the words as distributed representations then pass these to the LSTM and for each output we predict the next word.

For this we'll build a new dataloader with torchtext, the file we pass to the dataloader should contain one sentence per line, with words separated by whitespace.

```
word_1, ..., word_n
word_1, ..., word_k
...
```

in this dataloader you want to make sure that each sentence begins with a ```<start>``` token and ends with a ```<end>``` token, there is a keyword argument in ```Field``` for this :). But other than that, as before you read the dataset and output a iterator over the dataset and a vocabulary. 

Implement the dataloader, language model and the training loop (the training loop will basically be the same as for word2vec).

[12 marks]

In [8]:
# you can change these numbers to suit your needs as before :)
lm_hyperparameters = {'epochs':1,
                      'batch_size':4,
                      'learning_rate':0.001,
                      'embedding_dim':128,
                      'output_dim':128,
                      'hidden_dim': 128}

In [9]:
class LSTMDataset(Dataset):
    def __init__(self, data,
                 unk_token='UNK',
                 pad_token='PAD', vocab=False):            
        
        self.unk_token = unk_token
        self.pad_token = pad_token 
        
        self.words = set([unk_token, pad_token, "<start>", "<end>"]+[token.lower() for line in data for token in line])
        self.vocab = Vocab(self.words, pad_token) 
        self.pad_index = self.vocab.t2i[self.pad_token]
        self.unk_index = self.vocab.t2i[self.unk_token]
    
        self.examples = [[self.vocab.t2i[token] for token in line] 
                             for line in data]
        
            
    
    def __getitem__(self, idx):
        return self.examples[idx]

    def __len__(self):
        return len(self.examples)
    

In [10]:
data_path = '/Users/tomsodahlbladsjo/mlt/compSem/computational-semantics/assignments/03-distributed-representations-and-language-models/03-lab-data/wiki-corpus.50000.txt'
def get_data(path, batch_size):
    with open(path) as f:
        data = []
        for line in f:
            data.append(line.lower().split())
    dataset = LSTMDataset(data)
    dataloader = DataLoader(dataset,
                            batch_size=batch_size,
                            shuffle=True,
                            collate_fn=pad_fn)
    
    return dataloader, dataset.vocab

def pad_fn(data):
    lengths = [len(x) for x in data]
    max_len = max(lengths)
    padded_data = [[w for w in x]+[0]*(max_len-len(x)) for x in data]
    return padded_data

In [65]:
# Check that it is working and the format is right:
dataloader, vocab = get_data(data_path, 4)
print('pad_index =', vocab.t2i['PAD'])
for i, batch in enumerate(dataloader):
    print('Batch size:', len(batch)) 
    for example in batch:
        print(example)
        print('example length:', len(example))
  # Only look at first batch.
    break

pad_index = 0
Batch size: 4
[58520, 76012, 42319, 51369, 15780, 56434, 61880, 34677, 71720, 21261, 62970, 58520, 9463, 27568, 9463, 53175, 24250, 65229, 2338, 653, 4468, 2583, 51369, 4967, 76012, 72068, 74251, 50869, 16933, 9463, 75698, 9463, 26212, 65229, 21260, 30659, 48218]
example length: 37
[18445, 51073, 2777, 9463, 4967, 69986, 39501, 15780, 74505, 40780, 50869, 53795, 53533, 45404, 50869, 4967, 17757, 14485, 21261, 4967, 19641, 48218, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
example length: 37
[77112, 58940, 11438, 44189, 4967, 46858, 59436, 9463, 53175, 46548, 50182, 59092, 65541, 40822, 58506, 2583, 72101, 4967, 43111, 78585, 21261, 36490, 48930, 48218, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
example length: 37
[36623, 62983, 45992, 3258, 62285, 6994, 426, 41877, 48218, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
example length: 37


In [11]:
class LM_withLSTM(nn.Module):
    def __init__(self, vocab_dim, embedding_dim, output_dim, pad_index):
        super(LM_withLSTM, self).__init__()
        self.embeddings = nn.Embedding(num_embeddings=vocab_dim, 
                                               embedding_dim=embedding_dim,
                                               padding_idx=pad_index)
        self.LSTM = nn.LSTM(embedding_dim, output_dim)
        self.predict_word = nn.Linear(output_dim, vocab_dim)
    
    def forward(self, seq):
        embedded_seq = self.embeddings(seq)
        timestep_reprentation, *_ = self.LSTM(embedded_seq)
        predicted_words = self.predict_word(timestep_reprentation)
        
        return predicted_words

In [17]:
from tqdm.auto import tqdm

In [18]:
# load data
dataloader, vocab = get_data(data_path, lm_hyperparameters['batch_size'])

# build model and construct loss/optimizer
lm_model = LM_withLSTM(len(vocab)+1, 
                       lm_hyperparameters['embedding_dim'],
                       lm_hyperparameters['output_dim'],
                       vocab.t2i['PAD'])
lm_model.to(device)

loss_fn = nn.CrossEntropyLoss()
optimizer = optim.Adam(lm_model.parameters(), lr=lm_hyperparameters['learning_rate'])

# start training loop
total_loss = 0
for epoch in tqdm(range(lm_hyperparameters['epochs'])):
    print('Epoch:', epoch+1)
    for i, batch in enumerate(dataloader):
        
        # the strucure for each BATCH is:
        # <start>, w0, ..., wn, <end>
        
        # when training the model, at each input we predict the *NEXT* token
        # consequently there is nothing to predict when we give the model 
        # <end> as input. 
        # thus, we do not want to give <end> as input to the model, select 
        # from each batch all tokens except the last. 
        # tip: use pytorch indexing/slicing (same as numpy) 
        # (https://pytorch.org/tutorials/beginner/basics/tensorqs_tutorial.html#operations-on-tensors)
        # (https://jhui.github.io/2018/02/09/PyTorch-Basic-operations/)
        input_sentence = torch.Tensor([[vocab.t2i['<start>']]+sentence for sentence in batch]).long()
        
        # send your batch of sentences to the model
        output = lm_model(input_sentence)
        
        # for each output, the model predict the NEXT token, so we have to reshape 
        # our dataset again. On timestep t, we evaluate on token t+1. That is,
        # we never predict the <start> token ;) so this time, we select all but the first 
        # token from sentences (that is, all the tokens that we predict)
        gold_data = torch.Tensor([sentence+[vocab.t2i['<end>']] for sentence in batch]).long()
        
        # the shape of the output and sentence variable need to be changed,
        # for the loss function. Details are in the documentation.
        # You can use .view(...,...) to reshape the tensors  
        loss = loss_fn(output.view(lm_hyperparameters["batch_size"], len(vocab)+1, output.shape[1]), gold_data)
        total_loss += loss.item()
        
        # print average loss for the epoch
        print(total_loss/(i+1), end='\r') 
        
        # compute gradients
        loss.backward()
        
        # update parameters
        optimizer.step()
        
        # reset gradients
        optimizer.zero_grad()
    print()
    
# Pickle trained model:
pickle.dump(lm_model, open('lm_model.pickle', 'wb'))

  0%|          | 0/1 [00:00<?, ?it/s]

Epoch: 1
7.8163882299804685


In [72]:
# Load pickled model:
lm_model = pickle.load(open('lm_model.pickle', 'rb'))

(LM_withLSTM(
   (embeddings): Embedding(80718, 128, padding_idx=0)
   (LSTM): LSTM(128, 128)
   (predict_word): Linear(in_features=128, out_features=80718, bias=True)
 ),
 LM_withLSTM(
   (embeddings): Embedding(80718, 128, padding_idx=0)
   (LSTM): LSTM(128, 128)
   (predict_word): Linear(in_features=128, out_features=80718, bias=True)
 ))

### Evaluating the language model

We'll evaluate our model using the BLiMP dataset (https://github.com/alexwarstadt/blimp). The BLiMP dataset contains sets of linguistic minimal pairs for various syntactic and semantic phenomena, We'll evaluate our model on *existential quantifiers* (link: https://github.com/alexwarstadt/blimp/blob/master/data/existential_there_quantifiers_1.jsonl). This data, as the name suggests, investigate whether language models assign higher probability to *correct* usage of there-quantifiers. 

An example entry in the dataset is: 

```
{"sentence_good": "There was a documentary about music irritating Allison.", "sentence_bad": "There was each documentary about music irritating Allison.", "field": "semantics", "linguistics_term": "quantifiers", "UID": "existential_there_quantifiers_1", "simple_LM_method": true, "one_prefix_method": false, "two_prefix_method": false, "lexically_identical": false, "pairID": "0"}
```

Download the dataset and build a datareader (similar to what you did for word2vec). The dataset structure you should aim for is (you don't need to worry about the other keys for this assignment):

```
good_sentence_1, bad_sentence_1
...
```

your task now is to compare the probability assigned to the good sentence with to the probability assigned to the bad sentence. To compute a probability for a sentence we consider the product of the probabilities assigned to the *gold* tokens, remember, at timestep ```t``` we're predicting which token comes *next* e.g. ```t+1``` (basically, you do the same thing as you did when training).

In rough pseudo code what your code should do is:

```
accuracy = []
for good_sentence, bad_sentence in dataset:
    gs_lm_output = LanguageModel(good_sentence)
    gs_token_probabilities = softmax(gs_lm_output)
    gs_sentence_probability = product(gs_token_probabilities[GOLD_TOKENS])

    bs_lm_output = LanguageModel(bad_sentence)
    bs_token_probabilities = softmax(bs_lm_output)
    bs_sentence_probability = product(bs_token_probabilities[GOLD_TOKENS])

    # int(True) = 1 and int(False) = 0
    is_correct = int(gs_sentence_probability > bs_sentence_probability)
    accuracy.append(is_correct)

print(numpy.mean(accuracy))
    
```

[6 marks]

In [98]:
# your code goes here
import json
from operator import mul
from functools import reduce

def evaluate_model(path, vocab, model):
    
    accuracy = []
    with open(path) as f:
        # iterate over one pair of sentences at a time
        for line in f:
            # load the data
            data = json.loads(line)
            good_s = data['sentence_good']
            bad_s = data['sentence_bad']
            
            # the data is tokenized as whitespace
            tok_good_s = [token for token in good_s.lower().replace('.', ' .').replace(',', ' ,').split()]
            tok_bad_s = [token for token in bad_s.lower().replace('.', ' .').replace(',', ' ,').split()]
            
            # encode your words as integers using the vocab from the dataloader, size is (S)
            # we use unsqueeze to create the batch dimension 
            # in this case our input is only ONE batch, so the size of the tensor becomes: 
            # (S) -> (1, S) as the model expects batches
            enc_good_s = torch.tensor([vocab[x] if x in vocab.tokens else vocab['UNK'] for x in tok_good_s], device=device).unsqueeze(0)
            enc_bad_s = torch.tensor([vocab[x] if x in vocab.tokens else vocab['UNK'] for x in tok_bad_s], device=device).unsqueeze(0)

            # pass your encoded sentences to the model and predict the next tokens
            good_s = model(enc_good_s)
            bad_s = model(enc_bad_s)
            
            # get probabilities with softmax
            # Actually, softmax made the resulting probabilities so small that 
            # the sentence probabilities all just ended up as 0. 
            # We will use the logits instead.
            
            # gs_probs = F.softmax(good_s, dim=2).squeeze()
            # bs_probs = F.softmax(bad_s, dim=2).squeeze()

            # select the probability of the gold tokens
            gs_sent_prob = find_token_probs(good_s.squeeze(), enc_good_s)
            bs_sent_prob = find_token_probs(good_s.squeeze(), enc_bad_s)
            accuracy.append(int(gs_sent_prob>bs_sent_prob))
            
    return accuracy
            
def find_token_probs(model_probs, encoded_sentece):
    probs = []

    # iterate over the tokens in your encoded sentence
    for token, gold_token in enumerate(encoded_sentece.squeeze()):
        # select the probability of the gold tokens and save
        # hint: pytorch indexing is helpful here ;)
        prob = model_probs[token][gold_token]
        probs.append(prob)
        
    sentence_prob = reduce(mul, probs, 1)
    return sentence_prob

path = '03-lab-data/existential_there_quantifiers_1.jsonl'
accuracy = evaluate_model(path, vocab, lm_model)
print('Final accuracy:')
print(np.round(np.mean(accuracy), 3))


Final accuracy:
0.477


### Analysis

Our model get some score, say, 55% correct predictions. Is this good? Suggest some *baseline* (i.e. a stupid "model" we hope ours is better than) we can compare the model against.

[3 marks]

#### Answer:
Our model got a final accuracy score of 0.477. Since this is basically a binary task, where it either assigns a higher probability to the correct sentence than to the incorrect sentence or it doesn't, this result can be said to be slightly worse than chance. On the other hand, apart from the quantifiers the sentences in the dataset are very similar, so it is quite a difficult task for a language model.

Suggest some improvements you could make to your language model.

[3 marks]

#### Answer:
Again, more training time would hardly have hurt. We were also not entirely sure about where to insert the pad tokens in relation to the start and end tokens, so any implementation errors there (or anywhere else in the moddel architecture) might also have affected the model's performance.

Suggest some other metrics we can use to evaluate our system

[2 marks]

#### Answer:
An obvious evaluation approach would be to split the data into train and test, and basically do the same as during training. At every timestep, we could record the model prediction (that is, the most probable next token), and then measure the proportion of times that the model prediction == the actual next token of the sequence.

# Literature


Neural architectures:

[1] Y. Bengio, R. Ducharme, P. Vincent, and C. Janvin. A neural probabilistic language model. (Links to an external site.) Journal of Machine Learning Research, 3(6):1137–1155, 2003. (Sections 3 and 4 are less relevant today and hence you can glance through them quickly. Instead, look at the Mikolov papers where they describe training word embeddings with the current neural network architectures.)

[2] T. Mikolov, K. Chen, G. Corrado, and J. Dean. Efficient estimation of word representations in vector space. arXiv preprint arXiv:1301.3781, 2013.

[3] T. Mikolov, I. Sutskever, K. Chen, G. S. Corrado, and J. Dean. Distributed representations of words and phrases and their compositionality. In Advances in neural information processing systems, pages 3111–3119, 2013.
    


## Statement of contribution

Briefly state how many times you have met for discussions, who was present, to what degree each member contributed to the discussion and the final answers you are submitting.

-----
We all discussed the assignment continuously while working on it, both in person and in group chat. Most of the code in the final version was written by Tom, with additions and debugging by Katya, and with support from the rest of the group.

## Marks

This assignment has a total of 63 marks.